### PR TITLE
Issue  #10: support concatenation in yaml connectors for attribute mapping

### PR DESCRIPTION
**Pull request by kawtarBK9**
* Adjusted the concatenation in the MappingProcessor to invoke the concatenateColumnsValue method.
* Implement relevant test cases to reflect the corrected behavior.

### CODE REVIEW

**Review by NassimBtk — CHANGES_REQUESTED** [review]
Please update the integration test in the hardware module.
In `SuperConnectorCommon-header.yaml` define a new attribute for the `enclosure` monitor:
```yaml
attributes:
  vendor_and_model: $4 $5
   # other attributes...
```
Then in `expected.json`, add this new attribute (`vendor_and_model`) to the enclosure's expected attributes:
```json
    "attributes" : {
      "__display_id" : "000297800620 1",
      "vendor" : "EMC",
      "name" : "Enclosure: 000297800620 1 (EMC VMAX250F)",
      "connector_id" : "SuperConnectorOS",
      "model" : "VMAX250F",
      "id" : "2",
      "type" : "Storage",
      "info" : "Serial number : 297800620",
      "vendor_and_model" : "EMC VMAX250F"
    }
```

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessorTest.java [hunk 1]
```diff
@@ -687,4 +688,52 @@ void testInterpretNonContextMappingLookup() {
...
+		String value3 = "$1_$2_$3";
+		String key3 = "id";
+		String expectedResult3 = "DellOpenManage_cpu_1.1";
+		assertEquals(expectedResult3, mappingProcessor.concatenateColumnsValue(value3, key3));
+
+		// Test case where placeholder is beyond the size of the row
+		String invalidPlaceholder = "$4";
+		String key7 = "invalidPlaceholder";
+		String expectedResult7 = "";
+		assertEquals(expectedResult7, mappingProcessor.concatenateColumnsValue(invalidPlaceholder, key7));
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessorTest.java:719]: When the placeholder is beyond the row size we shouldn't force empty.

**Review by NassimBtk — CHANGES_REQUESTED** [review]
Some points to consider:

1. The regex is wrong and should be created as a constant. Here is the right pattern:
    ```java
      	private static final Pattern COLUMN_REFERENCE_PATTERN = Pattern.compile("\\$(\\d+)");
     ```
3. It is not required to use a _thread-safe_ object but even the entire object stringbuffer `resultBuffer` is not required in fact, to simplify and improve the code quality, we should invoke the [Matcher#replaceAll​(Function<MatchResult,​String> replacer)](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/regex/Matcher.html#replaceAll(java.util.function.Function)) introduced since Java 9.
4. The current implementation will fail if the column value contains a value such as `$N` where N is an number. Example: $1.
5. Javadoc should be completed and added even for private methods.
6. Integration tests should be added to SuperConnector IT.
7. Log messages need to be enhanced to report more context about failures.

Please consider the following implementation:

```java
	/**
	 * Creates a matcher that will match the given value against the column
	 * reference pattern.
	 *
	 * @param value The string value on which the regular expression pattern will be applied.
	 * @return A Matcher object that can be used to perform matching operations on the input string.
	 */
	private Matcher getColumnReferenceMatcher(String value) {
		return COLUMN_REFERENCE_PATTERN.matcher(value);
	}

	/**
	 * Replaces in the given value each column reference (E.g. $1) with the corresponding column value from the current row.
	 *
	 * @param value The input string containing placeholder notations.
	 * @param key   A key of the attribute.
	 * @return The modified string after replacing column references with actual values.
	 */
	protected String replaceColumnReferences(final String value, final String key) {
		return getColumnReferenceMatcher(value).replaceAll(match -> getColumnValue(match, key));
	}

	/**
	 * Retrieves the value of the column referenced in the match and replaces the match in the input string.
	 *
	 * @param match The MatchResult containing the column reference.
	 * @param key   A key of the attribute used for logging.
	 * @return The column value if available, or the original match if the column is not found.
	 */
	private String getColumnValue(final MatchResult match, final String key) {
		final int columnIndex = Integer.parseInt(match.group(1)) - 1;
		if (columnIndex >= 0 && columnIndex < row.size()) {
			final String columnValue =  row.get(columnIndex);
			if (columnValue != null) {
				return Matcher.quoteReplacement(columnValue);
			} else {
                               // TODO add context.
				log.warn(...);
			}
		} else {
                        // TODO add context. 
			log.warn(...);
		}

		return Matcher.quoteReplacement(match.group());
	}
```

**Review by NassimBtk — CHANGES_REQUESTED** [review]

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessor.java [hunk 2]
```diff
@@ -1091,6 +1129,15 @@ private boolean isColumnExtraction(String value) {
 		return getStringRegexMatcher(value).find();
 	}
 
+	/**
+	 * Checks whether the input string represents a column concatenation
```
[the remark this answers is not in the document]
Comment (TNohaic) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessor.java:1176]: JavaDoc still needs to be updated

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessor.java [hunk 3]
```diff
@@ -1091,6 +1129,15 @@ private boolean isColumnExtraction(String value) {
...
+	 * @param value
```
[the remark this answers is not in the document]
Comment (TNohaic) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessor.java:1177]: JavaDoc still needs to be updated

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessor.java [hunk 4]
```diff
@@ -1091,6 +1129,15 @@ private boolean isColumnExtraction(String value) {
...
+	 * @return true or false
```
[the remark this answers is not in the document]
Comment (TNohaic) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/utils/MappingProcessor.java:1178]: JavaDoc still needs to be updated

**Review by TNohaic — CHANGES_REQUESTED** [review]
Just a JavaDoc to update and it's good to go

**Review by TNohaic — APPROVED** [review]

**Review by NassimBtk — APPROVED** [review]

**Pull request merged**